node/src/components/linear_chain.rs

```diff
@@ -197,28 +271,65 @@ where
-                    .event(move |signature| Event::NewFinalitySignature(block_hash, signature)));
-                effects.extend(effect_builder.announce_block_added(block_hash, block_header).ignore());
+                    effect_builder
+                        .announce_block_added(block_hash, block_header)
+                        .ignore(),
+                );
                 effects
-            },
-            Event::NewFinalitySignature(block_hash, signature) => {
+            }
+            Event::NewFinalitySignature(fs) => {
+                let FinalitySignature {
+                    block_hash,
+                    signature,
+                    public_key,
+                } = *fs;
+                let verified =
+                    crypto::asymmetric_key::verify(block_hash.inner(), &signature, &public_key);
+                if let Err(err) = verified {
+                    error!(%block_hash, %public_key, %err, "received invalid finality signature");
+                    return Effects::new();
+                }
+                debug!(%block_hash, %public_key, "received new finality signature");
                 effect_builder
                     .get_block_from_storage(block_hash)
-                    .then(move |maybe_block| match maybe_block {
-                        Some(mut block) => {
-                            block.append_proof(signature);
-                            effect_builder.put_block_to_storage(Box::new(block))
-                        }
-                        None => {
-                            warn!("Received a signature for {} but block was not found in the Linear chain storage", block_hash);
-                            panic!("Unhandled")
-                        }
+                    .event(move |maybe_block| {
+                        Event::BlockSignatureContinuation(Box::new(maybe_block), fs)
                     })
-                    .ignore()
+            }
+            Event::BlockSignatureContinuation(maybe_block, fs) => match *maybe_block {
+                None => {
+                    let block_hash = fs.block_hash();
+                    let public_key = fs.public_key();
+                    warn!(%block_hash, %public_key, "received a signature for a block that was not found in storage");
+                    Effects::new()
+                }
+                Some(mut block) => {
+                    if !block.contains_proof(fs.signature()) {
+                        block.append_proof(fs.signature);
+                        effect_builder
+                            .put_block_to_storage(Box::new(block))
+                            .event(move |_| Event::GossipFinalitySignature(fs))
+                    } else {
+                        Effects::new()
+                    }
+                }
             },
+            Event::GossipFinalitySignature(fs) => effect_builder
+                .broadcast_message(Message::FinalitySignature(*fs))
+                .ignore(),
```

Comment:
I am not a fan of this split and introduction of new event variants but I couldn't get the future-based version to compile even due to problems with integrating `Future` and  `Effects` types.